## Normalize TPM to the ERCC spike ins
Xulong Wang (xulong.wang@jax.org)

***

```{r, echo = FALSE, warning = FALSE}
library(ape)
library(amap)
library(ggplot2)
library(ggdendro)

rm(list = ls())
```

##### Gene expression in transcript and gene levels were estimated with RSEM. Genes with bigger than 5 TPM in at least 2 samples were kept for downstream analysis 

```{r}
setwd("~/Dropbox/GitHub/FY")
load("tpm.rdt")

data <- data.frame(row.names = tpm.dt$Symbol, tpm.dt[2:ncol(tpm.dt)])
data <- data[apply(data, 1, function (x) length(x[x>5]) > 2), ]

dim(data)
```

##### Separate the ERCC spike ins from the others
```{r}
genes <- data[! grepl("ERCC", rownames(data)), ]
spike <- data[grep("ERCC", rownames(data)), ]
dim(genes)
dim(spike)
```

##### ERCC mol concentration vs TPM plot
```{r}
mol <- read.delim("cms_095046.txt", stringsAsFactors = F)
mol <- mol[match(rownames(spike), mol$ERCC.ID), 4] * 0.02

par(mfrow = c(3, 6))
for (i in 1:ncol(spike)) plot(log2(mol), log2(spike[, i]), main = colnames(spike)[i])
```

##### Normalization among samples with scaling factors from modeling the ERCC transcripts with a linear model
```{r}
sf <- apply(spike[, 2:ncol(spike)], 2, function (x) summary(lm(x ~ -1 + spike[, 1]))$coefficients[1, 1])
sf <- c(M1IN = 1, sf)
genes <- t(apply(genes, 1, "/", sf))
```

##### Generate datasets of each gene's mean expression value over the replicates, together with the relative values to the IN samples
```{r}
uid <- gsub("[123]", "", colnames(genes))
geno <- gsub("^(W|M).*", "\\1", uid)
cond <- c("WIN", "MIN", "WNONP", "MNONP", "WPLM", "MPLM") 

fc <- NULL
for (idx in cond) fc <- cbind(fc, rowMeans(genes[, uid == idx]))
colnames(fc) <- cond

fc <- cbind(fc, WNONP2WIN = fc[, "WNONP"] / fc[, "WIN"], MNONP2MIN = fc[, "MNONP"] / fc[, "MIN"],
                WPLM2WIN = fc[, "WPLM"] / fc[, "WIN"], MPLM2MIN = fc[, "MPLM"] / fc[, "MIN"])

tail(fc)
```

##### A two way hierarchical clustering on the samples
```{r}
hc1 <- hcluster(t(genes), method = "pearson", link = "average")
plot(as.phylo(hc1), edge.width = 2, direction = "downward")
```